core/capabilities/capability.go

```diff
@@ -0,0 +1,145 @@
+package capabilities
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"golang.org/x/mod/semver"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+// CapabilityType is an enum for the type of capability.
+type CapabilityType string
+
+// CapabilityType enum values.
+const (
+	CapabilityTypeTrigger CapabilityType = "trigger"
+	CapabilityTypeAction  CapabilityType = "action"
+	CapabilityTypeReport  CapabilityType = "report"
+	CapabilityTypeTarget  CapabilityType = "target"
+)
+
+// IsValid checks if the capability type is valid.
+func (c CapabilityType) IsValid() error {
+	switch c {
+	case CapabilityTypeTrigger,
+		CapabilityTypeAction,
+		CapabilityTypeReport,
+		CapabilityTypeTarget:
+		return nil
+	}
+
+	return fmt.Errorf("invalid capability type: %s", c)
+}
+
+// Validatable is an interface for validating the config and inputs of a capability.
+type Validatable interface {
+	ValidateConfig(config values.Map) error
+	ExampleOutput() values.Value
+	ValidateInput(inputs values.Map) error
+}
+
+// Executable is an interface for executing a capability.
+type Executable interface {
+	// Start will be called when the capability is loaded by the application.
+	// Start will be called before the capability is added to the registry.
+	Start(ctx context.Context, config values.Map) (values.Value, error)
+	// Capability must respect context.Done and cleanup any request specific resources
+	// when the context is cancelled. When a request has been completed the capability
+	// is also expected to close the callback channel.
+	// Request specific configuration is passed in via the inputs parameter.
+	Execute(ctx context.Context, callback chan values.Value, inputs values.Map) error
+	// Stop will be called before the application exits.
+	// Stop will be called after the capability is removed from the registry.
+	Stop(ctx context.Context) error
+}
+
+// Capability is an interface for a capability.
+type Capability interface {
+	Executable
+	Validatable
+	Info() CapabilityInfo
+}
+
+// CapabilityInfo is a struct for the info of a capability.
+type CapabilityInfo struct {
+	ID             string
+	CapabilityType CapabilityType
+	Description    string
+	Version        string
+}
+
+// Info returns the info of the capability.
+func (c CapabilityInfo) Info() CapabilityInfo {
```

Comment:
May want to return map[string]CapabilityInfo so we can add name to execute.

Reply:
This was a separate conversation that is resolved now - we've parked this for now and @cedric-cordenier will write an exploration doc on whether we want to pass in a name field or not.